File: kevoree-core/org.kevoree.framework/src/main/scala/org/kevoree/framework/aspects/ContainerNodeAspect.scala
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kevoree.framework.aspects

import org.kevoree._
import KevoreeAspects._
import org.slf4j.LoggerFactory
import collection.mutable.ListBuffer
import scala.Some
import scala.collection.JavaConversions._

case class ContainerNodeAspect(node: ContainerNode) {
  private val logger = LoggerFactory.getLogger(this.getClass)

  def isModelEquals(ct: ContainerNode): Boolean = {
    ct.getName == node.getName
    /* TODO deep compare */
  }

  def getComponentTypes: List[ComponentType] = {
    var alreadyDeployComponentType: List[ComponentType] = List()
    node.getComponents.foreach {
      c =>
        if (!alreadyDeployComponentType.exists({
          e => e.getName == c.getTypeDefinition.getName
        })) {
          alreadyDeployComponentType = alreadyDeployComponentType ++
            List(c.getTypeDefinition.asInstanceOf[ComponentType])
        }
    }
    alreadyDeployComponentType
  }

  def getUsedTypeDefinition: List[TypeDefinition] = {
    var usedType: Set[TypeDefinition] = Set()

    /* ADD NODE TYPE DEFINITION */
    usedType = usedType ++ getTypeAndInherited(node.getTypeDefinition)
    /* ADD COMPONENT TYPE USED */
    node.getComponents.foreach(c => usedType = usedType ++ getTypeAndInherited(c.getTypeDefinition))

    /* ADD CHANNEL TYPE USED */
    /* add channel fragment on node */
    node.eContainer.asInstanceOf[ContainerRoot].getMBindings.foreach {
      mb =>
        if (mb.getPort.eContainer.eContainer == node) {
          if (!usedType.exists({
            e => e.getName == mb.getHub.getTypeDefinition.getName
          })) {
            usedType = usedType ++ getTypeAndInherited(mb.getHub.getTypeDefinition)
          }
        }
    }

    /* add group type on node */
    /* add group */
    node.eContainer.asInstanceOf[ContainerRoot].getGroups.filter(group => group.getSubNodes.contains(node)).foreach({
      c =>
        usedType = usedType ++ getTypeAndInherited(c.getTypeDefinition)

    })

    usedType.toList
  }

  def getChannelFragment: List[Channel] = {
    /* add channel fragment on node */
    val usedChannel: ListBuffer[Channel] = ListBuffer[Channel]()
    node.getComponents.foreach {
      component =>
        (component.getProvided.toList ++ component.getRequired.toList).foreach {
          port => port.getBindings.foreach {
            mbinding => if (!usedChannel.exists(c => c.getName == mbinding.getHub.getName)) {
              usedChannel += mbinding.getHub
            }
          }
        }
    }
    usedChannel.toList
  }

  def getGroups: List[Group] = {
    var usedGroup: List[Group] = List()
    node.eContainer.asInstanceOf[ContainerRoot].getGroups.filter(group => group.getSubNodes.contains(node))
      .foreach(group => {
      usedGroup = usedGroup ++ List(group)
    })

    usedGroup
  }


  def getInstances: List[Instance] = getGroups ++ getChannelFragment ++ node.getComponents

  private def getTypeAndInherited(t: TypeDefinition): List[TypeDefinition] = {
    var types = List[TypeDefinition]()
    if (t.getSuperTypes != null) {
      t.getSuperTypes.foreach {
        superT =>
          types = types ++ getTypeAndInherited(superT)
      }
      types = types ++ List(t)
    }
    types
  }

  def isDeployUnitUsed(du: DeployUnit): Boolean = {
    node.getUsedTypeDefinition.exists(usedTypeDef => {
      val usedDU = usedTypeDef.foundRelevantDeployUnit(node)
      usedDU.isDeployUnitUsed(du)
    })
  }


  def getKevoreeVersion: String = {
    try {

      val rDU = node.getTypeDefinition.foundRelevantDeployUnit(node)
      if (rDU != null) {
        rDU.getRequiredLibs.find(du => du.getGroupName == "org.kevoree" && du.getUnitName == "org.kevoree.api") match {
          case None => {
            logger.error("Error found api deploy unit for " + node.getName)
            ""
          } // must never appear
          case Some(du) => du.getVersion
        }

      } else {
        logger.error("Error found relevant deploy unit for " + node.getName)
        ""
      }
    } catch {
      case _@e => logger.debug("Unable to find kevoree version", e); ""
    }
  }


}

File: kevoree-core/org.kevoree.framework/src/main/java/org/kevoree/framework/kaspects/ContainerNodeAspect.kt
package org.kevoree.framework.kaspects

import java.util.ArrayList
import java.util.HashSet
import org.kevoree.Channel
import org.kevoree.ContainerNode
import org.kevoree.ContainerRoot
import org.kevoree.Group
import org.kevoree.TypeDefinition

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 10/03/13
 * Time: 09:49
 */
class ContainerNodeAspect {

    private fun getTypeAndInherited(t: TypeDefinition?): List<TypeDefinition> {
        var types = ArrayList<TypeDefinition>()
        if (t != null && t.superTypes != null) {
            for(superT in t.superTypes){
                types.addAll(getTypeAndInherited(superT))
            }
            types.add(t)
        }
        return types
    }

    fun getUsedTypeDefinition(node: ContainerNode): List<TypeDefinition> {
        var usedType: HashSet<TypeDefinition> = HashSet()
        /* ADD NODE TYPE DEFINITION */
        usedType.addAll(getTypeAndInherited(node.typeDefinition))
        /* ADD COMPONENT TYPE USED */
        for(c in node.components){
            usedType.addAll(getTypeAndInherited(c.typeDefinition))
        }

        /* ADD CHANNEL TYPE USED */
        /* add channel fragment on node */
        val rootModel = node.eContainer() as ContainerRoot
        for(mb in rootModel.mBindings) {
            if (mb.port!!.eContainer()!!.eContainer() == node) {
                usedType.addAll(getTypeAndInherited(mb.hub!!.typeDefinition))
            }
        }

        /* add group type on node */
        /* add group */
        for(group in rootModel.groups){
            if(group.subNodes.contains(node)){
                val groupTypeDef = group.typeDefinition
                if(groupTypeDef != null){
                    usedType.add(groupTypeDef)
                }
            }
        }
        return usedType.toList()
    }

    fun getChannelFragment(node: ContainerNode): List<Channel> {
        /* add channel fragment on node */
        val usedChannels = ArrayList<Channel>()
        for (component in node.components) {
            for (port in component.provided) {
                for (mb in port.bindings) {
                    if (!usedChannels.contains(mb.hub)) {
                        usedChannels.add(mb.hub!!)
                    }
                }
            }
            for (port in component.required) {
                for (mb in port.bindings) {
                    if (!usedChannels.contains(mb.hub)) {
                        usedChannels.add(mb.hub!!)
                    }
                }
            }
        }
        return usedChannels
    }

    fun getGroups(node: ContainerNode): List<Group> {
        val usedGroups = ArrayList<Group>()
        for (group in (node.eContainer() as ContainerRoot).groups) {
            if (group.subNodes.contains(node)) {
                usedGroups.add(group)
            }
        }
        return usedGroups
    }

}